File: Shortest_Path_Visiting_All_Nodes.py
class Solution:

    def shortestPathLength(self, graph: List[List[int]]) -> int:

        n = len(graph)

        arr = deque([(i, 1 << i, 0) for i in range(n)])

        past = set([(i, 1 << i) for i in range(n)])

        ans = 0

        while arr:

            node, path, le = arr.popleft()

            if path == (1 << n) - 1:

                ans = le

                break

            for i in graph[node]:

                mask_ = path | (1 << i)

                if (i, mask_) not in past:

                    arr.append((i, mask_, le + 1))

                    past.add((i, mask_))

        return ans

File: Shortest_Path_Visiting_All_Nodes.js
var shortestPathLength = function (graph) {

    let l = graph.length

    let arr = []

    for (let i = 0; i < l; i++) {
        arr.push([i, 1 << i, 0])
    }

    let past = new Set()

    let ans = 0

    while (arr.length !== 0) {

        // console.log(arr)

        const [node, path, cnt] = arr.shift()

        if (path === (1 << l) - 1) {
            ans = cnt
            break
        }

        for (let i = 0; i < graph[node].length; i++) {

            let mask_ = path | (1 << graph[node][i])

            if (!past.has(`${graph[node][i]}${mask_}`)) {
                past.add(`${graph[node][i]}${mask_}`)
                arr.push([graph[node][i], mask_, cnt + 1])
            }
        }

    }

    return ans

};